server/market/market.go

```diff
@@ -3,30 +3,442 @@
+	defer epoch.ordersMtx.Unlock()
+
+	// Verify that the order is not already in the epoch queue
+	ord := rec.order
+	if epoch.Orders[ord.ID()] != nil {
+		// TODO: remove this check when we are sure the order router is behaving
+		log.Errorf("Received duplicate order %v!", ord)
+		errChan <- ErrDuplicateOrder
+		return
+	}
+
+	// Validate the order. The order router must do it's own validation, but do
+	// a second validation for (1) this Market and (2) epoch status, before
+	// putting it on the queue.
+	if !m.validateOrder(ord) {
+		log.Errorf("(*Market).runEpochs: Invalid order received for epoch %d: %v",
+			epoch.Epoch, ord)
+		errChan <- ErrInvalidOrder
+		return
+	}
+
+	// Sign the order and prepare the client response. Only after the archiver
+	// has successfully stored the new epoch order should the order be committed
+	// for processing.
+	respMsg, err := m.orderResponse(rec, uint64(epoch.Epoch))
+	if err != nil {
+		log.Errorf("failed to create msgjson.Message for order %v, msgID %v response: %v",
+			rec.order, rec.msgID, err)
+		errChan <- ErrMalformedOrderResponse
+		return
+	}
+
+	// Archive the new epoch order BEFORE inserting it into the epoch queue,
+	// initiating the swap, and notifying book subscribers.
+	if err := m.storage.NewEpochOrder(ord); err != nil {
+		log.Errorf("(*Market).runEpochs: Failed to store new epoch order %v: %v",
+			ord, err)
+		errChan <- ErrInternalServer
+		m.cancel()
+		return
+	}
+
+	// Insert the order into the epoch queue.
+	epoch.Orders[ord.ID()] = ord
+	errChan <- nil
+
+	// Inform the client that the order has been received, stamped, signed, and
+	// inserted into the current epoch queue.
+	m.auth.Send(ord.User(), respMsg)
+
+	// Send epoch update to order book subscribers.
+	go m.notifyNewOrder(ord, epochAction)
+}
+
+// processEpoch performs the following operations for a closed epoch, which must
+// no longer be modified by another goroutine:
+//  1. Perform matching with the order book.
+//  2. Send book and unbook notifications to the book subscribers.
+//  3. Initiate the swap negotiation via the Market's Swapper.
+func (m *Market) processEpoch(epoch *EpochQueue) {
+	// Perform the matching. The matcher updates the order book.
+	orders := epoch.OrderSlice()
+	m.bookMtx.Lock()
+	matches, _, failed, partial, booked, unbooked := m.matcher.Match(m.book, orders)
+	m.bookMtx.Unlock()
+	log.Debugf("Matching complete for epoch %d:"+
+		" %d matches (%d partial fills), %d booked, %d unbooked, %d failed",
+		epoch.Epoch, len(matches), len(partial), len(booked), len(unbooked), len(failed),
+	)
+	// TODO get and store important epoch data.
+
+	// Send "book" notifications to order book subscribers.
+	for _, ord := range booked {
+		m.notifyNewOrder(ord, bookAction)
+	}
+
+	// Send "unbook" notifications to order book subscribers.
+	for _, ord := range unbooked {
+		m.notifyNewOrder(ord, unbookAction)
+	}
+
+	// Exclude cancel orders from the matches sent to the swap coordinator.
```

Comment:
Why are these being excluded? The swap coordinator sends a notification to these clients.